server/src/main/java/org/elasticsearch/cluster/coordination/NodeRemovalClusterStateTaskExecutor.java

```diff
@@ -22,14 +23,16 @@
     private final AllocationService allocationService;
     private final Logger logger;
 
-    public static class Task {
+    public static class Task implements ClusterStateTaskListener {
 
         private final DiscoveryNode node;
         private final String reason;
+        private final ClusterStateTaskListener listener;
 
-        public Task(final DiscoveryNode node, final String reason) {
+        public Task(DiscoveryNode node, String reason, ClusterStateTaskListener listener) {
```

Comment:
Rather than passing in a whole `ClusterStateTaskListener` we could just pass in a `Runnable` that calls into the `JoinReasonService` on success and make the logging calls ourselves.

Strangely this executor uses the logger for the `Coordinator` which is kind of confusing. Let's give it its own `private static final Logger logger` instead.